build_feature_tree.js

```diff
@@ -0,0 +1,143 @@
+                foundKeys.push(objPath);
+            } else if (typeof obj[k] === "object" && obj[k] !== null) {
+                iterate(obj[k], objPath);
+            }
+        });
+    };
+    iterate(object);
+    return foundKeys;
+};
+
+/**
+ * Patched version of lodash's _.get() which handles empty object paths.
+ * @param {Object} obj - The object to query
+ * @param {string} objPath - The path of the property to get
+ * @returns {*} - Resolved value
+ * @todo Move to code.js
+ */
+const getByPath = (obj, objPath) => {
+    return !objPath ? obj : _.get(obj, objPath);
+};
+
+/**
+ * Load file and resolve references recursively.
+ * NOTE: The `include` key is used to designate a reference.
+ * With a separator present, the reference may refer to a key in the object.
+ * Please use with caution in order to avoid circular references!
+ * @param filePath
+ * @param separator
+ * @returns {Object}
+ * @todo Move to code.js
+ */
+const loadYAMLWithReferences = (filePath, separator = "#") => {
+    const fileContent = fs.readFileSync(filePath, "utf8");
+    const currentDir = path.dirname(filePath);
+    let yamlObj = yaml.load(fileContent);
+    findKeys("include", yamlObj).forEach((includePath) => {
+        console.log(`found include at path ${includePath}`);
+        const reference = _.get(yamlObj, includePath);
+        const parentObjPath = includePath.replace(/[.]?include/, "");
+        const currObj = getByPath(yamlObj, parentObjPath);
+        if (typeof reference === "string") {
+            const [fileName, key = ""] = reference.split(separator);
+            let resolved = loadYAMLWithReferences(path.resolve(currentDir, fileName));
+            resolved = getByPath(resolved, key);
+            resolved = utils.recursiveMerge(resolved, currObj);
+            if (!parentObjPath) {
+                yamlObj = resolved;
+            } else {
+                _.set(yamlObj, parentObjPath, resolved);
+            }
+            _.unset(yamlObj, includePath);
+        }
+        return null;
+    });
+    return yamlObj;
+};
+
+const cleanObjectPath = (objPath) => {
+    if (!objPath.includes(".")) return objPath;
+    return `['${objPath}']`;
+};
+
+/**
+ * Load YAML asset files and insert asset data into the MODEL_DATA object.
+ * @param {string} dir - folder containing asset file
+ * @param {string} fileName - asset file name
+ * @param {string} assetExtension - file extension for asset
+ */
+const loadAssetFile = (dir, fileName, assetExtension = ".yml") => {
+    const yamlObj = loadYAMLWithReferences(path.resolve(dir, fileName));
+    const key = cleanObjectPath(path.basename(fileName, assetExtension));
+    let objectPath = path.relative(ASSET_PATH, dir).split(path.sep).join(".");
+    objectPath = [objectPath, key].filter(Boolean).join(".");
+    _.set(FEATURE_DATA, objectPath, yamlObj);
+    console.log(`setting feature data of [${fileName}] at path [${objectPath}]`);
+};
+
+const getDirectories = (currentPath) => {
```

Comment:
Seems like a fair amount of code duplicate between here and https://github.com/Exabyte-io/mode.js/pull/11/files#diff-50f84a390329578d5c1edb8fe724d2234bd932ed64726b9f93d46621ae6c0364R20. Can we avoid it somehow?

Reply:
yes, there already is a task to refactor functions related to parsing asset files.